ts-client/src/vault/index.ts

```diff
@@ -0,0 +1,330 @@
+
+import { VaultDetails, VaultImplementation, VaultProgram, VaultState } from "./types";
+import { deserializeAccount, getAssociatedTokenAccount, getLpSupply, getOnchainTime, getOrCreateATAInstruction, getVaultPdas, unwrapSOLInstruction, wrapSOLInstruction } from "./utils";
+import { LOCKED_PROFIT_DEGRADATION_DENOMINATOR, PROGRAM_ID, SOL_MINT, VAULT_STRATEGY_ADDRESS } from "./constants";
+import { getStrategyHandler, getStrategyType, StrategyState } from "./strategy";
+import { IDL, Vault as VaultIdl } from "./idl";
+
+const getVaultState = async (
+    vaultParams: TokenInfo,
+    program: VaultProgram
+): Promise<{ vaultPda: PublicKey, tokenVaultPda: PublicKey, vaultState: VaultState, lpSupply: string }> => {
+    const { vaultPda, tokenVaultPda } = await getVaultPdas(new PublicKey(vaultParams.address), new PublicKey(program.programId));
+    const vaultState = (await program.account.vault.fetchNullable(
+        vaultPda
+    )) as VaultState;
+    const lpSupply = await getLpSupply(program.provider.connection, vaultState.lpMint);
+
+    if (!vaultState) {
+        throw "Cannot get vault state";
+    }
+    return { vaultPda, tokenVaultPda, vaultState, lpSupply };
+}
+
+const getVaultLiquidity = async (connection: Connection, tokenVaultPda: PublicKey): Promise<string | null> => {
+    const vaultLiquidityResponse = await connection.getAccountInfo(tokenVaultPda)
+    if (!vaultLiquidityResponse) return null;
+
+    const vaultLiquidtySerialize = deserializeAccount(vaultLiquidityResponse.data)
+    return vaultLiquidtySerialize?.amount.toString() || null;
+}
+
+export default class VaultImpl implements VaultImplementation {
+    private connection: Connection;
+    private cluster: Cluster = 'mainnet-beta';
+
+    // Vault
+    private tokenInfo: TokenInfo;
+    private program: VaultProgram;
+
+    public vaultPda: PublicKey;
+    public tokenVaultPda: PublicKey;
+    public vaultState: VaultState;
+    public lpSupply: string = '';
+
+    private constructor(program: VaultProgram, vaultDetails: VaultDetails, opt?: { cluster?: Cluster }) {
+        this.connection = program.provider.connection;
+        this.cluster = opt?.cluster ?? 'mainnet-beta';
+
+        this.tokenInfo = vaultDetails.vaultParams;
+        this.program = program;
+        this.vaultPda = vaultDetails.vaultPda;
+        this.tokenVaultPda = vaultDetails.tokenVaultPda;
+        this.vaultState = vaultDetails.vaultState;
+        this.lpSupply = vaultDetails.lpSupply;
+    }
+
+    public static async create(
+        connection: Connection,
+        vaultParams: TokenInfo,
+        opt?: {
+            cluster?: Cluster,
+            programId?: string,
+        }
+    ): Promise<VaultImpl> {
+        const provider = new AnchorProvider(connection, {} as any, AnchorProvider.defaultOptions());
+        const program = new Program<VaultIdl>(
+            IDL as VaultIdl,
+            opt?.programId || PROGRAM_ID,
+            provider
+        );
```

Comment:
I realised if we create new program here, every vault we're gonna create 1 program instance. We should only share 1 program instance

Reply:
I gave this a thought earlier, but I don't want our integrator have to care about `IDL` file

Reply:
we can cache the program instance, if the program doesn't exists only we initiate it